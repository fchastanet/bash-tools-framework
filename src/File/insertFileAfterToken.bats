#!/usr/bin/env bash

# shellcheck source=src/batsHeaders.sh
source "$(cd "${BATS_TEST_DIRNAME}/.." && pwd)/batsHeaders.sh"

# shellcheck source=/src/File/insertFileAfterToken.sh
source "${srcDir}/File/insertFileAfterToken.sh"

function File::insertFileAfterToken::OneToken { #@test
  cp "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken.txt" \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken.txt"
  File::insertFileAfterToken \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken.toInject.txt" \
    "# FUNCTIONS"

  diff "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken.expected.txt" \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken.txt"
}

function File::insertFileAfterToken::TwoTokens { #@test
  cp "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken2.txt" \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken2.txt"
  File::insertFileAfterToken \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken2.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken.toInject.txt" \
    "# FUNCTIONS"

  diff "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken2.expected.txt" \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken2.txt" >&3
}

function File::insertFileAfterToken::InsertTwice { #@test
  cp "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken2.txt" \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken3.txt"
  File::insertFileAfterToken \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken3.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken.toInject.txt" \
    "# FUNCTIONS"
  File::insertFileAfterToken \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken3.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken.toInject3.txt" \
    "# FUNCTIONS"

  diff "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken3.expected.txt" \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken3.txt"
}

function File::insertFileAfterToken::NoMatchingPattern { #@test
  cp "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken2.txt" \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken4.txt"
  run File::insertFileAfterToken \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken4.txt" \
    "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken.toInject.txt" \
    "# NOT MATCHING PATTERN"
  assert_success
  diff "${BATS_TEST_DIRNAME}/testsData/insertFileAfterToken2.txt" \
    "${BATS_TEST_TMPDIR}/insertFileAfterToken4.txt"
}
