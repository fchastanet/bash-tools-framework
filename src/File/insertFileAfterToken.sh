#!/usr/bin/env bash

# Public: insert file content inside another file after a
# pattern
#
# @param {String} file $1 file in which fileToImport will be
#   injected
# @param {String} fileToImport $2 file to inject before the
#   token
# @param {String} token $3 token needs to be properly escaped
#   for usage inside sed // regexp
File::insertFileAfterToken() {
  local file="$1"
  local fileToImport="$2"
  local token="$3"

  # /${token}/! if line does not match pattern
  #    b command is applied : Branch unconditionally to label.
  #       The label may be omitted, in which case the next cycle
  #       is started.
  #    else
  #       r ${fileToImport} : loads the file in pattern space
  #       n;i \\${NL}: n prints the pattern space and add a new line
  #       :a;n;ba : this allows to iterate until the end of the file
  #          :a label a
  #          n: print pattern space (current line)
  #          ba: goto label a
  # n command: replace the pattern space with the next line of input.
  #   If there is no more input then sed exits without processing any
  #   more commands.

  # trick to insert a new line with sed, as sed does not support \n
  local NL=$'\n'
  sed -i -E -e $"
      /${token}/!b
      r ${fileToImport}
      n;i \\${NL}
      :a;n;ba
    " "${file}"
}
