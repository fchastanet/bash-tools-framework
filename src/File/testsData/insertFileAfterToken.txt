stuff before
# FUNCTIONS
stuff after
