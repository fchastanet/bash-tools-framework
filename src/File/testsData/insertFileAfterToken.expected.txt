stuff before
# FUNCTIONS
insertFileAfterToken injected

stuff after
